Update the readme with floating point (#497)
wasmx/fizzy: README.md

- [ ] Easily embeddable (*and take part of the standardisation of the "C/C++ embedding API"*)

II) Simplicity
- [x] Only implement WebAssembly 1.0 (and none of the proposals)
- [x] Interpreter only
- [x] Support only WebAssembly binary encoding as an input (no support for the text format (`.wat`/`.wast`))

III) Conformance

> - [x] Should have 99+% unit test coverage

Reviewer: I'd say we aim for 100% although it may be difficult to get precisely that because of the toolchain bugs and limitations.

Author: That's why I put 99%, but I'm fine either way.

Reviewer 2: "99+%" looks good to me

This is the unit tests suite of Fizzy.

## Special notes about floating point

### Exceptions

Certain floating point operations can emit *exceptions* as defined by the IEEE 754 standard.

> (Here is a [summary from glibc](https://www.gnu.org/software/libc/manual/html_node/FP-Exceptions.html)).

Reviewer: ```suggestion
(Here is a [summary from GNU C Library](https://www.gnu.org/software/libc/manual/html_node/FP-Exceptions.html)).
```

It is however up to the language how these manifest, and in C/C++ depending on the settings they can result in
a) setting of a flag; b) in *traps*, such as the `SIGFPE` signal.

Fizzy does not manipulate this setting, but expects it to be left at option a) of above, which is the default.

> In glibc this can be controlled via the [`feenableexcept` and `fedisableexcept` functions](https://www.gnu.org/savannah-checkouts/gnu/libc/manual/html_node/Control-Functions.html).

Reviewer: ```suggestion
In GNU C Library this can be controlled via the [`feenableexcept` and `fedisableexcept` functions](https://www.gnu.org/savannah-checkouts/gnu/libc/manual/html_node/Control-Functions.html).
```

> The caller has to ensure this is not set to *trap* mode. We strongly suggest not to enable it and catch the signal,

Reviewer 2: Nitpick: maybe it's only me, but maybe it could be unclear whether we suggest to catch the signal or not to catch
```suggestion
The caller has to ensure this is not set to *trap* mode. We strongly suggest to avoid enabling it and catching the signal,
```

Reviewer: For me the second sentence is also confusing. Maybe just "The behavior with traps enabled is unpredictable/not guaranteed/...".

> The caller has to ensure this is not set to *trap* mode. We strongly suggest not to enable it and catch the signal,

Reviewer: ```suggestion
The user of the Fizzy library has to ensure this is not set to *trap* mode. We strongly suggest not to enable it and catch the signal,
```

because in that case correct behaviour can not be guaranteed.

### Rounding

> The IEEE 754 standard also defines four different rounding directions (or modes):

Reviewer: ```suggestion
The IEEE 754 standard defines four rounding directions (or modes):
```

> - to nearest, tie-to-even,

Reviewer: ```suggestion
- to nearest, tie to even (default)
```

> - downward (towards -inf),

Reviewer 2: maybe better
```suggestion
- downward (towards -infinity),
```

> - downward (towards -inf),

Reviewer: ```suggestion
- toward -∞ (rounding down, floor)
```

> - upward (towards inf),

Reviewer: ```suggestion
- toward +∞ (rounding up, ceiling)
```

> - towards zero (truncate).

Reviewer: ```suggestion
- toward  0 (truncation)
```

> - toward -∞ (rounding down, floor),

Author: ```suggestion
- toward -∞ (rounding down aka floor),
- toward +∞ (rounding up aka ceiling),
```

@chfast how about this?

Reviewer: Both "rounding down" and "floor" are aka. Just drop "floor" and "ceiling".

Author: Well the current one was your suggestion: https://github.com/wasmx/fizzy/pull/497#discussion_r473072120 :wink:

Reviewer: I took it from wikipedia, but it is more readable without the last alias.

> WebAssembly requires "to nearest" rounding, which is the first option, and is the default.

Reviewer: ```suggestion
The semantic of the WebAssembly floating-point instructions is based on the default "no nearest" rounding direction. Therefore, Fizzy expects this rounding direction to be selected. Otherwise, the result of some instructions may be different than what the WebAssembly specification predicts, i.e. the behavior will follow the selected rounding direction.
```

The list of these instructions will be provided later when we have tests in place.

The `floor` instruction should be not on the list, that's why we are having the workaround for the GCC bug already in place.

Reviewer 2: To me this `floor` hack is an inconsistency and I think it should be mentioned, if we want to keep it.

Reviewer 2: Now I see how the hack is consistent with keeping the instructions affected by rounding mode limited to a logical set. Ok to add a list of affected instructions later.

In the GNU C Library the rounding mode can be controlled via the [`fesetround` and `fegetround` functions](https://www.gnu.org/software/libc/manual/html_node/Rounding.html).

> Fizzy does not manipulate this setting, but expects it to be left at the default, otherwise correct behaviour can not be guaranteed.

Reviewer: ```suggestion

```
Should be drop if the above change is accepted.
